### PR TITLE
Drop `request.timer` attribute.

### PR DESCRIPTION
**Pull request by tomchristie**
Minor bit of refactoring. We don't really want a `request.timer` attribute in our public API, exposing an internal `ElapsedTimer` instance.

Instead support a `Response(..., stream=..., get_elapsed_time=...)` parameter, which allows passing in a callable that is used to set `response.elapsed` when the response is closed. We don't particularly expect our users to need to use this, but it's a little neater and more tightly scoped than having an accessible timer on the *request* instance.

Also allows us to use a backend-native timer, appropriate to whichever context we're running in.

### CODE REVIEW

File: tests/models/test_responses.py [hunk 1]
```diff
@@ -31,7 +30,6 @@ def test_response():
     assert response.text == "Hello, world!"
     assert response.request.method == "GET"
     assert response.request.url == "https://example.org"
-    assert response.elapsed >= datetime.timedelta(0)
```
Comment (florimondmanca) [inline tests/models/test_responses.py:34]: Haven't checked, do we still have some tests for the `response.elapsed` property?

**Review by florimondmanca — APPROVED** [review]
Nice bit of uncoupling. Left a sanity check question but this looks good!

File: tests/models/test_responses.py (hunk 1 above)
Comment (tomchristie) [reply tests/models/test_responses.py:34]: Yup `test_client.py` and `test_async_client.py` check the property, and `test_responses.py` has a test to ensure that it's not accessible until the response has been closed.

**Review by florimondmanca — DISMISSED** [review]

**Comment by tomchristie** [comment]
Before we merge this, do we think we've got the right name?
Some options might be...

* `get_elapsed_time=...` *Currently using, in this PR*
* `elapsed_timer=...`
* `elapsed_callable=...`
* `elapsed_time=...`
* `elapsed_func=...`

Anything else? Which of those is preferable?

**Comment by florimondmanca** [comment]
> `elapsed_func=...`

Yup, I think I prefer this one better. (True that `.get_elapsed_time` looks like a method rather than an externally-passed argument.)

**Pull request merged**